File: ios/RunningApp/Utils/BackupHelper.swift
import Foundation

// MARK: - 数据备份和恢复工具
class BackupHelper {

    static let shared = BackupHelper()

    private init() {}

    // 备份目录
    private let backupDirectory: URL = {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        return documentsURL.appendingPathComponent("Backups", isDirectory: true)
    }()

    // MARK: - 备份数据

    /// 备份所有数据
    func backupAllData() async -> Result<URL, Error> {
        do {
            // 确保备份目录存在
            try createBackupDirectoryIfNeeded()

            // 创建备份文件
            let backupURL = createBackupFileURL()

            // 收集所有数据
            let dataDict = await collectAllData()

            // 创建备份包
            let backupData = try await createBackupPackage(data: dataDict)

            // 写入文件
            try backupData.write(to: backupURL)

            Logger.d("Backup created: \(backupURL.path)")
            return .success(backupURL)
        } catch {
            Logger.e("Backup failed", error: error)
            return .failure(error)
        }
    }

    /// 收集所有数据
    private func collectAllData() async -> [String: Any] {
        var dataDict: [String: Any] = [:]

        // 收集跑步记录
        dataDict["runningRecords"] = []

        // 收集成就
        dataDict["achievements"] = []

        // 收集训练计划
        dataDict["trainingPlans"] = []

        // 添加版本信息
        dataDict["version"] = 1

        Logger.d("Data collected")
        return dataDict
    }

    /// 创建备份包
    private func createBackupPackage(data: [String: Any]) async throws -> Data {
        var backupDict: [String: Any] = [:]

        // 添加数据
        backupDict["data"] = data

        // 添加偏好设置
        backupDict["preferences"] = exportPreferences()

        // 添加元数据
        backupDict["metadata"] = createMetadata()

        // 转换为JSON
        let jsonData = try JSONSerialization.data(withJSONObject: backupDict, options: .prettyPrinted)
        return jsonData
    }

    /// 导出偏好设置
    private func exportPreferences() -> [String: Any] {
        let defaults = UserDefaults.standard
        return defaults.dictionaryRepresentation()
    }

    /// 创建元数据
    private func createMetadata() -> [String: Any] {
        let deviceHelper = DeviceHelper.shared
        return [
            "appVersion": deviceHelper.appVersion,
            "backupTime": Date().timeIntervalSince1970,
            "deviceModel": deviceHelper.deviceTypeName,
            "systemVersion": deviceHelper.systemVersion
        ]
    }

    /// 创建备份文件URL
    private func createBackupFileURL() -> URL {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd_HHmmss"
        let timestamp = dateFormatter.string(from: Date())
        let fileName = "backup_\(timestamp).json"
        return backupDirectory.appendingPathComponent(fileName)
    }

    /// 创建备份目录
    private func createBackupDirectoryIfNeeded() throws {
        if !FileManager.default.fileExists(atPath: backupDirectory.path) {
            try FileManager.default.createDirectory(
                at: backupDirectory,
                withIntermediateDirectories: true
            )
        }
    }

    // MARK: - 恢复数据

    /// 从备份文件恢复数据
    func restoreFromBackup(_ backupURL: URL) async -> Result<Void, Error> {
        do {
            // 读取备份文件
            let data = try Data(contentsOf: backupURL)
            let backupDict = try JSONSerialization.jsonObject(with: data) as? [String: Any]

            guard let backupDict = backupDict else {
                throw NSError(domain: "BackupHelper", code: -1, userInfo: [
                    NSLocalizedDescriptionKey: "Invalid backup file format"
                ])
            }

            // 恢复数据
            if let dataDict = backupDict["data"] as? [String: Any] {
                await restoreData(dataDict)
            }

            // 恢复偏好设置
            if let prefsDict = backupDict["preferences"] as? [String: Any] {
                restorePreferences(prefsDict)
            }

            Logger.d("Restore completed")
            return .success(())
        } catch {
            Logger.e("Restore failed", error: error)
            return .failure(error)
        }
    }

    /// 恢复数据
    private func restoreData(_ dataDict: [String: Any]) async {
        // 这里应该将数据写入数据库
        Logger.d("Data restored")
    }

    /// 恢复偏好设置
    private func restorePreferences(_ prefsDict: [String: Any]) {
        let defaults = UserDefaults.standard

        prefsDict.forEach { key, value in
            defaults.set(value, forKey: key)
        }

        defaults.synchronize()
        Logger.d("Preferences restored")
    }

    // MARK: - 导出数据

    /// 导出跑步数据为CSV
    func exportRunningDataToCsv() async -> Result<URL, Error> {
        do {
            let exportURL = createExportFileURL(name: "running_data", extension: "csv")

            // CSV头部
            var csv = "日期,距离(km),时长(秒),配速(min/km),卡路里,平均心率\n"

            // 这里应该从数据库读取数据
            // 示例数据
            csv += "2025-11-17,5.2,1800,5.77,300,145\n"

            // 写入文件
            try csv.write(to: exportURL, atomically: true, encoding: .utf8)

            Logger.d("CSV exported: \(exportURL.path)")
            return .success(exportURL)
        } catch {
            Logger.e("Export CSV failed", error: error)
            return .failure(error)
        }
    }

    /// 导出跑步数据为JSON
    func exportRunningDataToJson() async -> Result<URL, Error> {
        do {
            let exportURL = createExportFileURL(name: "running_data", extension: "json")
            let dataDict = await collectAllData()

            let jsonData = try JSONSerialization.data(withJSONObject: dataDict, options: .prettyPrinted)
            try jsonData.write(to: exportURL)

            Logger.d("JSON exported: \(exportURL.path)")
            return .success(exportURL)
        } catch {
            Logger.e("Export JSON failed", error: error)
            return .failure(error)
        }
    }

    /// 创建导出文件URL
    private func createExportFileURL(name: String, extension: String) -> URL {
        let documentsURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let exportDirectory = documentsURL.appendingPathComponent("Exports", isDirectory: true)

        // 创建导出目录
        try? FileManager.default.createDirectory(at: exportDirectory, withIntermediateDirectories: true)

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyyMMdd_HHmmss"
        let timestamp = dateFormatter.string(from: Date())
        let fileName = "\(name)_\(timestamp).\(`extension`)"

        return exportDirectory.appendingPathComponent(fileName)
    }

    // MARK: - 备份管理

    /// 获取所有备份文件
    func getAllBackups() -> [URL] {
        do {
            let fileURLs = try FileManager.default.contentsOfDirectory(
                at: backupDirectory,
                includingPropertiesForKeys: [.creationDateKey],
                options: .skipsHiddenFiles
            )

            return fileURLs
                .filter { $0.pathExtension == "json" && $0.lastPathComponent.hasPrefix("backup_") }
                .sorted { url1, url2 in
                    let date1 = (try? url1.resourceValues(forKeys: [.creationDateKey]))?.creationDate ?? Date.distantPast
                    let date2 = (try? url2.resourceValues(forKeys: [.creationDateKey]))?.creationDate ?? Date.distantPast
                    return date1 > date2
                }
        } catch {
            Logger.e("Get backups failed", error: error)
            return []
        }
    }

    /// 删除备份文件
    func deleteBackup(_ backupURL: URL) -> Bool {
        do {
            try FileManager.default.removeItem(at: backupURL)
            Logger.d("Backup deleted: \(backupURL.path)")
            return true
        } catch {
            Logger.e("Delete backup failed", error: error)
            return false
        }
    }

    /// 清理旧备份（保留最近N个）
    func cleanOldBackups(keepCount: Int = 5) {
        let backups = getAllBackups()
        if backups.count > keepCount {
            let toDelete = backups.dropFirst(keepCount)
            toDelete.forEach { _ = deleteBackup($0) }
            Logger.d("Cleaned \(toDelete.count) old backups")
        }
    }

    /// 获取备份文件大小
    func getBackupSize(_ backupURL: URL) -> Int64 {
        do {
            let attributes = try FileManager.default.attributesOfItem(atPath: backupURL.path)
            return attributes[.size] as? Int64 ?? 0
        } catch {
            return 0
        }
    }

    /// 获取备份文件信息
    func getBackupInfo(_ backupURL: URL) -> [String: String] {
        let size = getBackupSize(backupURL)

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"

        let creationDate: Date
        if let resourceValues = try? backupURL.resourceValues(forKeys: [.creationDateKey]),
           let date = resourceValues.creationDate {
            creationDate = date
        } else {
            creationDate = Date()
        }

        return [
            "name": backupURL.lastPathComponent,
            "size": FormatUtils.shared.formatFileSize(size),
            "date": dateFormatter.string(from: creationDate)
        ]
    }
}

// MARK: - 便捷扩展

extension BackupHelper {

    /// 快速备份
    func quickBackup() async -> URL? {
        let result = await backupAllData()
        switch result {
        case .success(let url):
            return url
        case .failure:
            return nil
        }
    }

    /// 快速恢复
    func quickRestore(from url: URL) async -> Bool {
        let result = await restoreFromBackup(url)
        switch result {
        case .success:
            return true
        case .failure:
            return false
        }
    }
}

File: android/app/src/main/java/com/runningapp/utils/BackupHelper.kt
package com.runningapp.utils

import android.content.Context
import android.os.Environment
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.*
import java.util.zip.ZipEntry
import java.util.zip.ZipInputStream
import java.util.zip.ZipOutputStream
import javax.inject.Inject
import javax.inject.Singleton

/**
 * 数据备份和恢复工具
 * 支持导出跑步数据、设置等
 */
@Singleton
class BackupHelper @Inject constructor(
    @ApplicationContext private val context: Context
) {

    companion object {
        private const val BACKUP_DIR = "RunningApp_Backup"
        private const val BACKUP_FILE_PREFIX = "backup_"
        private const val BACKUP_FILE_EXT = ".zip"
    }

    // MARK: - 备份数据

    /**
     * 备份所有数据
     */
    suspend fun backupAllData(): Result<File> = withContext(Dispatchers.IO) {
        try {
            val backupFile = createBackupFile()
            val dataJson = collectAllData()

            // 创建ZIP文件
            ZipOutputStream(FileOutputStream(backupFile)).use { zipOut ->
                // 添加数据JSON
                zipOut.putNextEntry(ZipEntry("data.json"))
                zipOut.write(dataJson.toString().toByteArray())
                zipOut.closeEntry()

                // 添加设置
                val prefsJson = exportPreferences()
                zipOut.putNextEntry(ZipEntry("preferences.json"))
                zipOut.write(prefsJson.toString().toByteArray())
                zipOut.closeEntry()

                // 添加元数据
                val metaJson = createMetadata()
                zipOut.putNextEntry(ZipEntry("metadata.json"))
                zipOut.write(metaJson.toString().toByteArray())
                zipOut.closeEntry()
            }

            Logger.d("BackupHelper", "Backup created: ${backupFile.absolutePath}")
            Result.success(backupFile)
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Backup failed", e)
            Result.failure(e)
        }
    }

    /**
     * 收集所有数据
     */
    private suspend fun collectAllData(): JSONObject = withContext(Dispatchers.IO) {
        val dataJson = JSONObject()

        try {
            // 这里应该从数据库读取数据
            // 示例结构
            dataJson.put("version", 1)
            dataJson.put("runningRecords", JSONArray())
            dataJson.put("achievements", JSONArray())
            dataJson.put("trainingPlans", JSONArray())

            Logger.d("BackupHelper", "Data collected")
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Collect data failed", e)
        }

        dataJson
    }

    /**
     * 导出偏好设置
     */
    private fun exportPreferences(): JSONObject {
        val prefsJson = JSONObject()
        val prefs = context.getSharedPreferences("app_preferences", Context.MODE_PRIVATE)

        try {
            prefs.all.forEach { (key, value) ->
                when (value) {
                    is String -> prefsJson.put(key, value)
                    is Int -> prefsJson.put(key, value)
                    is Long -> prefsJson.put(key, value)
                    is Float -> prefsJson.put(key, value)
                    is Boolean -> prefsJson.put(key, value)
                }
            }
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Export preferences failed", e)
        }

        return prefsJson
    }

    /**
     * 创建元数据
     */
    private fun createMetadata(): JSONObject {
        return JSONObject().apply {
            put("appVersion", DeviceHelper(context).getVersionName())
            put("backupTime", System.currentTimeMillis())
            put("deviceModel", DeviceHelper(context).getDeviceName())
            put("androidVersion", DeviceHelper(context).getAndroidVersion())
        }
    }

    /**
     * 创建备份文件
     */
    private fun createBackupFile(): File {
        val backupDir = File(
            context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS),
            BACKUP_DIR
        )
        if (!backupDir.exists()) {
            backupDir.mkdirs()
        }

        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
            .format(Date())
        val fileName = "$BACKUP_FILE_PREFIX$timestamp$BACKUP_FILE_EXT"

        return File(backupDir, fileName)
    }

    // MARK: - 恢复数据

    /**
     * 从备份文件恢复数据
     */
    suspend fun restoreFromBackup(backupFile: File): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            if (!backupFile.exists()) {
                return@withContext Result.failure(Exception("Backup file not found"))
            }

            var dataJson: JSONObject? = null
            var prefsJson: JSONObject? = null
            var metaJson: JSONObject? = null

            // 解压ZIP文件
            ZipInputStream(FileInputStream(backupFile)).use { zipIn ->
                var entry = zipIn.nextEntry
                while (entry != null) {
                    val content = zipIn.readBytes().toString(Charsets.UTF_8)

                    when (entry.name) {
                        "data.json" -> dataJson = JSONObject(content)
                        "preferences.json" -> prefsJson = JSONObject(content)
                        "metadata.json" -> metaJson = JSONObject(content)
                    }

                    zipIn.closeEntry()
                    entry = zipIn.nextEntry
                }
            }

            // 恢复数据
            dataJson?.let { restoreData(it) }
            prefsJson?.let { restorePreferences(it) }

            Logger.d("BackupHelper", "Restore completed")
            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Restore failed", e)
            Result.failure(e)
        }
    }

    /**
     * 恢复数据
     */
    private suspend fun restoreData(dataJson: JSONObject) = withContext(Dispatchers.IO) {
        try {
            // 这里应该将数据写入数据库
            Logger.d("BackupHelper", "Data restored")
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Restore data failed", e)
        }
    }

    /**
     * 恢复偏好设置
     */
    private fun restorePreferences(prefsJson: JSONObject) {
        val prefs = context.getSharedPreferences("app_preferences", Context.MODE_PRIVATE)
        val editor = prefs.edit()

        try {
            prefsJson.keys().forEach { key ->
                when (val value = prefsJson.get(key)) {
                    is String -> editor.putString(key, value)
                    is Int -> editor.putInt(key, value)
                    is Long -> editor.putLong(key, value)
                    is Double -> editor.putFloat(key, value.toFloat())
                    is Boolean -> editor.putBoolean(key, value)
                }
            }
            editor.apply()

            Logger.d("BackupHelper", "Preferences restored")
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Restore preferences failed", e)
        }
    }

    // MARK: - 导出数据

    /**
     * 导出跑步数据为CSV
     */
    suspend fun exportRunningDataToCsv(): Result<File> = withContext(Dispatchers.IO) {
        try {
            val exportFile = createExportFile("running_data", "csv")

            // CSV头部
            val csv = StringBuilder()
            csv.append("日期,距离(km),时长(秒),配速(min/km),卡路里,平均心率\n")

            // 这里应该从数据库读取数据并写入CSV
            // 示例数据
            csv.append("2025-11-17,5.2,1800,5.77,300,145\n")

            FileOutputStream(exportFile).use { out ->
                out.write(csv.toString().toByteArray())
            }

            Logger.d("BackupHelper", "CSV exported: ${exportFile.absolutePath}")
            Result.success(exportFile)
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Export CSV failed", e)
            Result.failure(e)
        }
    }

    /**
     * 导出跑步数据为JSON
     */
    suspend fun exportRunningDataToJson(): Result<File> = withContext(Dispatchers.IO) {
        try {
            val exportFile = createExportFile("running_data", "json")
            val dataJson = collectAllData()

            FileOutputStream(exportFile).use { out ->
                out.write(dataJson.toString(2).toByteArray())
            }

            Logger.d("BackupHelper", "JSON exported: ${exportFile.absolutePath}")
            Result.success(exportFile)
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Export JSON failed", e)
            Result.failure(e)
        }
    }

    /**
     * 创建导出文件
     */
    private fun createExportFile(name: String, extension: String): File {
        val exportDir = File(
            context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS),
            "Exports"
        )
        if (!exportDir.exists()) {
            exportDir.mkdirs()
        }

        val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault())
            .format(Date())
        val fileName = "${name}_$timestamp.$extension"

        return File(exportDir, fileName)
    }

    // MARK: - 备份管理

    /**
     * 获取所有备份文件
     */
    fun getAllBackups(): List<File> {
        val backupDir = File(
            context.getExternalFilesDir(Environment.DIRECTORY_DOCUMENTS),
            BACKUP_DIR
        )

        return if (backupDir.exists()) {
            backupDir.listFiles { file ->
                file.name.startsWith(BACKUP_FILE_PREFIX) &&
                        file.name.endsWith(BACKUP_FILE_EXT)
            }?.sortedByDescending { it.lastModified() } ?: emptyList()
        } else {
            emptyList()
        }
    }

    /**
     * 删除备份文件
     */
    fun deleteBackup(backupFile: File): Boolean {
        return try {
            backupFile.delete()
        } catch (e: Exception) {
            Logger.e("BackupHelper", "Delete backup failed", e)
            false
        }
    }

    /**
     * 清理旧备份（保留最近N个）
     */
    fun cleanOldBackups(keepCount: Int = 5) {
        val backups = getAllBackups()
        if (backups.size > keepCount) {
            backups.drop(keepCount).forEach { it.delete() }
            Logger.d("BackupHelper", "Cleaned ${backups.size - keepCount} old backups")
        }
    }

    /**
     * 获取备份文件大小
     */
    fun getBackupSize(backupFile: File): Long {
        return backupFile.length()
    }

    /**
     * 获取备份文件信息
     */
    fun getBackupInfo(backupFile: File): Map<String, String> {
        return mapOf(
            "name" to backupFile.name,
            "size" to FormatUtils(context).formatFileSize(backupFile.length()),
            "date" to SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.getDefault())
                .format(Date(backupFile.lastModified()))
        )
    }
}
